proposals/p5434.md

```diff
@@ -0,0 +1,1331 @@
+
+fn ParensInBraces(...)
+    -> {-> .a: bool,
+        -> .b: (->let f32, ->var C),
+        ->ref .c: i32};
+
+let {a: bool,
+     .b = (x: f32, var y: C),
+     ref c: i32} = ParensInBraces(...);
+```
+
+This feature is intended to support cases like `enumerate` that will want to
+return a value for the index but a reference to the element of the sequence
+being enumerated.
+
+### Nested binding patterns
+
+Since a `ref` binding may only bind to a durable reference expression, it can't
+be used to bind the result of a function returning an initializing expression.
+However, if the initializing expression is bound to a `var`, any nested patterns
+are reference binding patterns bound to the subobject, following
+[proposal #5164: "Updates to pattern matching for objects"](https://github.com/carbon-language/carbon-lang/pull/5164).
+
+For example:
+
+```carbon
+fn F() -> (bool, (C, i32));
+let (b: bool, var (c: C, i: i32)) = F();
+```
+
+is equivalent to:
+
+```carbon
+fn F() -> (bool, (C, i32));
+let (b: bool, var v: (C, i32)) = F();
+let ref c: C = v.0;
+let ref i: i32 = v.1;
+```
+
+### `bound` parameters
+
+It is an error to return a reference to a temporary object that won't live once
+the function returns, even if it is a parameter marked `bound`.
+
+```carbon
+fn Invalid1() -> i32* {
+  var x: i32 = 4;
+  // ❌ Invalid.
+  return &x;
+}
+
+fn Invalid2(bound x: i32) -> ref i32 {
+  var y: i32 = x;
+  // ❌ Invalid.
+  return y;
+}
+
+// ✅ Valid
+fn Valid1(bound p: i32*) -> ref i32 {
+  return *p;
+}
+
+fn Invalid3(bound var x: i32) -> ref i32 {
+  // ❌ Invalid: lifetime of `var` parameter
+  // ends when function returns.
+  return x;
+}
+
+class ReturnMember {
+  // ✅ Valid
+  fn ValidRef[bound ref self: Self]() -> ref i32 {
+    return self.m;
+  }
+
+  // ❌ Invalid: can't return reference to value.
+  fn InvalidLet[bound self: Self]() -> ref i32 {
+    return self.m;
+  }
+
+  // ❌ Invalid: `var self` lifetime ends.
+  fn InvalidVar[bound var self: Self]()
+      -> ref i32 { return self.m; }
+
+  var m: i32;
+}
+
+class DerefPointerMember {
+  // ✅ Valid
+  fn ValidRef[bound ref self: Self]() -> ref i32 {
+    return *self.pm;
+  }
+
+  // ✅ Valid
+  fn ValidLet[bound self: Self]() -> ref i32 {
+    return *self.pm;
+  }
+
+  // ✅ Valid
+  fn ValidVar[bound var self: Self]()
+      -> ref i32 { return *self.pm; }
+
+  var pm: i32*;
+}
+```
+
+The `bound` parameters must include everything referenced by the return, but
+need not be referenced, particularly not on every code path.
+
+```carbon
+// Result references `r` if `b` is true, and `p`
+// otherwise. Valid as long as both are marked `bound`.
+fn Conditional(b: bool, bound ref r: C, bound p: C*)
+    -> ref C {
+  if (b) {
+    return r;
+  } else {
+    return *p;
+  }
+}
+```
+
+The parameters of functions defined in an interface may also be marked as
+`bound`. The `impl` of that interface for a type can only have `bound` in a
+subset of the positions present in the interface.
```

Comment:
The "can only have `bound` in a subset" phrasing is a little hard to follow. How about this?

```suggestion
`bound`. The `impl` of that interface for a type can omit occurrences of
`bound` from the interface, but cannot add new ones.
```